### DISCUSSION TITLE
List of primitive Atoms && find || filter

### DISCUSSION THREAD

**Discussion by bwoutersvaa** [pmndrs/jotai#662] [Q&A]
We got an application with a list of invoices. We already make an atom<PrimitiveAtom<Invoice>[]>();
So in the application we want to filter the list of primtive atoms. But it's a hard job, because the value of useAtom isn't an invoice, but a PrimitiveAtom<Invoice>. 

To explain the flow:
1. User visits overview
2. Get's list of invoices
3. Sets the primitiveInvoicesAtom
4. Clicks on one atom
5. Set the active atom

Now comes the ceveat.
The user can also visit a detail page directly. In that case the list isn't filled in with data. 

```
export const primitiveInvoicesAtom = atom<PrimitiveAtom<Invoice>[]>([]);
export const activeInvoiceIdAtom = atom<string | undefined>(undefined);

// This gives a warning because we cannot read invoiceid from a primitiveatom
export const activeInvoice = atom(
  (get) =>
    get(primitiveInvoicesAtom).find(
      (invoice) => invoice.invoiceId.toString() === get(activeInvoiceIdAtom),
    )
);

```

So in short i want the following:
1. A List of invoice atoms (Pure for update reasons)
2. Get the active invoice atom from the list based on the id via props or url. (derived)
3. At best return a writable atom because we need to update the invoice based on actions.

I think array mutation based on primitive atoms is still a bit unclear when reading the docs if the atom is not passed in by props.

**Comment by dai-shi** [comment]
Hi, thanks for opening a discussion.

In short:
```js
export const activeInvoice = atom(
  (get) =>
    get(primitiveInvoicesAtom).find(
      (invoice) => get(invoice).invoiceId.toString() === get(activeInvoiceIdAtom),
    )
);
```

I know it's a little bit hard to reach this mental model. It's probably a doc issue. Suggestions are welcome.

> 2. Get the active invoice atom from the list based on the id via props or url. (derived)

If the id is coming from props, you want to create a derived atom (non-global) in component.

```js
const Component = ({ id }) => {
  const activeInvoice = useMemo(() => atom(
  (get) =>
    get(primitiveInvoicesAtom).find(
      (invoice) => get(invoice).invoiceId.toString() === id,
    )
  ), [id])
  // continued
```

> 3. At best return a writable atom because we need to update the invoice based on actions.

You can then use the atom in the same component.

```js
  const [invoice, setInvoice] = useAtom(activeInvoice)
```

**Reply by bwoutersvaa** [reply]
Thanks for the quick reply!

I get it that it's better to use a not global atom, but when i try it that the types aren't really helping.

```
const primitiveInvoicesAtom = atom<PrimitiveAtom<Invoice>[]>([]);

const Component = ({ id }) => {
  const activeInvoice = useMemo(() => atom(
  (get) =>
    get(primitiveInvoicesAtom).find(
      (invoice) => get(invoice).invoiceId.toString() === id,
    )
  ), [id])
  // continued

  const [invoice, setInvoice] = useAtom(activeInvoice);
}
```

The return type is of Atom<PrimitiveAtom<Invoice>>. When using it in a useAtom it only gets past the < Atom >.
So the invoice is of type <PrimitiveAtom<Invoice>> and isn't read so still can't use the values. 

setInvoice cannot be used because it is of type never. Need to make a custom setter?


I like the way the atom with atoms work, to keep the update cycle really small. Because we only have to update the single atom, when using the atoms as props.

**Reply by bwoutersvaa** [reply]
Maybe to explain it a bit, we got a page with a list on the left side, and a detail view on the right side. They share the same data, except the detail view is a single invoice, instead of a list. So when we change the invoice in the detail screen, i want the overview to be updated based on the same atom. That's why we use a PrimitiveAtom[]. The row component in the list uses a useAtom with the primitiveAtom based on a prop atom.

**Reply by bwoutersvaa** [reply]
Got a bit further and got a working GETTER example (typings). Have to check if the state will update like we want to:
Have to check if an atom was found and we have to use get(PrimitiveAtom<Invoice>);
    return foundAtom && get(foundAtom);

This way we can use it in the useAtom function. 

SETTER is still an issue though

**Reply by dai-shi** [reply]
> setInvoice cannot be used because it is of type never. 

That's strange. Can you create a tiny codesandbox to play with?

**Reply by bwoutersvaa** [reply]
Thanks for your help so far! 

https://codesandbox.io/s/intelligent-franklin-1qf1r

**Reply by dai-shi** [reply]
sorry, there was a bug in my original comment.
here's the fix.
https://codesandbox.io/s/exciting-shadow-51n45

---
https://github.com/pmndrs/jotai/discussions/662#discussioncomment-1189656
The correct snippet is:
```js
const Component = ({ id }) => {
  const activeInvoice = useAtom(useMemo(() => atom(
  (get) =>
    get(primitiveInvoicesAtom).find(
      (invoice) => get(invoice).invoiceId.toString() === id,
    )
  ), [id]))
  // continued
```

**Reply by bwoutersvaa** [reply]
Thanks man! Really appreciate the help! Problem with my code was the return get().

IT's not possible to use the Getter outside an atom function? Like some sort of utility method?

**Reply by dai-shi** [reply]
No, as you guessed. Getter must be inside, so that we can track the dependency.
